### DISCUSSION TITLE
Unsubscribe to event

### DISCUSSION THREAD

**Discussion by paromarc** [AxisCommunications/acap3-examples#172] [Q&A] [answered]
Hi, I am running the "subscribe_to_event" example and when I try to unsubscribe using 
ax_event_handler_unsubscribe(event_handler, subscription, NULL);
ax_event_handler_free(event_handler);

Basically I am subscribing in the main and unsubscribing right after.

I am getting this from the terminal:

root@axis-b8a44f050c2b:/usr/local/packages/subscribe_to_event# ./subscribe_to_event
** Message: 10:48:27.537: Unsubscribe to event

(process:8319): GLib-CRITICAL **: 10:48:29.549: g_hash_table_remove_internal: assertion 'hash_table != NULL' failed

(process:8319): GLib-CRITICAL **: 10:48:29.549: g_hash_table_destroy: assertion 'hash_table != NULL' failed

Any idea what that means? Tested on Q1615 MKIII fw 10.12.73

Thanks,
Marco

**Comment by pataxis** [comment, marked as answer]
Hi  @paromarc and thanks for your question.

Do you get this error also when building and running the "subscribe_to_event" example without any modifications?

**Reply by paromarc** [reply]
Hi @pataxis ! Thanks for your quick reply :-)
No, I don't as the example doesn't include the code to unsubscribe to said event.
This is the only modification I have made to test it out:
// Event handler
  event_handler = ax_event_handler_new();
  subscription = onviftrigger_subscription(event_handler, &token);

  g_message("Unsubscribe to event");
  ax_event_handler_unsubscribe(event_handler, subscription, NULL);
  ax_event_handler_free(event_handler);

I noticed if I remove "ax_event_handler_free(event_handler);" the message goes away.
My only concern is a potential memory leak if I don't free the handler, or is it taken care when unsubscribing ?

Thanks!
Marco

**Reply by pataxis** [reply]
I'm afraid I can't reproduce your error. I tested to build the example unmodified, SSH in to the device and then
```sh
cd /usr/local/packages/subscribe_to_event
./subscribe_to_event
```
and got no output in the terminal.

I'm not sure what changes you have made in your local example. Have you removed the `g_main_loop_run` and added a `g_message`? In the `subscribe_to_event` example [main loop](https://github.com/AxisCommunications/acap3-examples/blob/main/axevent/subscribe_to_event/app/subscribe_to_event.c#L114-L144) the same cleanup functions you refer to are called
```c
  // Cleanup event handler
  ax_event_handler_unsubscribe(event_handler, subscription, NULL);
  ax_event_handler_free(event_handler);
```

**Reply by paromarc** [reply]
My main goal is being able to unsubscribe to an existing event and subscribe to a new one. Before I do that I call the unsubscribe and free functions (and that is when I get the error). What is the recommended way of doing so? Thanks

**Reply by pataxis** [reply]
The recommended way is to create an `AxEventHandler` with `ax_event_handler_new()` once and reuse it throughout the program.
There is no need to free the handler after an unsubscribe and there should not be any memory leak.
The handler can be freed at the end of the program `ax_event_handler_free()`.

**Reply by paromarc** [reply]
Perfect! That's how I have it right now in my app. My only concern was a memory leak if not freeing the event, but if that is not the case I am good with it :-)

Thanks a lot for your help on this one!
Marco
